packages/kit/src/runtime/client/client.js

```diff
@@ -2089,8 +2089,11 @@ function _start_router() {
 
 		if (download) return;
 
+		const [nonhash, hash] = url.href.split('#');
+		const same_pathname = nonhash === strip_hash(location);
+
 		// Ignore the following but fire beforeNavigate
-		if (external || options.reload) {
+		if (external || (options.reload && !same_pathname && !hash)) {
```

Comment:
Why `&& !hash`? Wouldn't this mean you're not getting a reload if you're linking to a page within your app and include a hash?
```suggestion
		if (external || (options.reload && !same_pathname)) {
```

Reply:
Yes, that's what we're trying to avoid. Currently, there is no reload, so instead of letting it behave as if there will be a reload, we want to let the hash link handling take over instead.

Reply:
The current logic would not do a full page reload if you would navigate from `/foo` to `/bar#baz`, because there's a hash. That sounds wrong to me.

I gotta refine my suggestion, I think we need the reload if the pathname is different, or if there's no hash (most browser reload the page in that case). Only if there's a hash we don't want to reload.

```suggestion
		if (external || (options.reload && (!same_pathname || !hash))) {
```